src/cls/queue/cls_rgw_queue.cc

```diff
@@ -0,0 +1,664 @@
+    decode(can_fit, iter);
+    CLS_LOG(1, "INFO: Can urgent data fit: %d \n", can_fit);
+
+    //insert as xattrs
+    if (! can_fit) {
+      //remove inserted entry from urgent data
+      urgent_data.urgent_data_map.erase(op.info.tag);
+      urgent_data.num_head_urgent_entries -= 1;
+      has_urgent_data = false;
+
+      bufferlist bl_xattrs;
+      int ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl_xattrs);
+      if (ret < 0 && (ret != -ENOENT && ret != -ENODATA)) {
+        CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+        return ret;
+      }
+      std::unordered_map<string,ceph::real_time> xattr_urgent_data_map;
+      if (ret != -ENOENT && ret != -ENODATA) {
+        auto iter = bl_xattrs.cbegin();
+        try {
+          decode(xattr_urgent_data_map, iter);
+        } catch (buffer::error& err) {
+          CLS_LOG(1, "ERROR: cls_gc_queue_remove(): failed to decode xattrs urgent data map\n");
+          return -EINVAL;
+        } //end - catch
+      }
+      xattr_urgent_data_map.insert({op.info.tag, op.info.time});
+      urgent_data.num_xattr_urgent_entries += 1;
+      has_urgent_data = true;
+      bufferlist bl_map;
+      encode(xattr_urgent_data_map, bl_map);
+      ret = cls_cxx_setxattr(hctx, "cls_queue_urgent_data", &bl_map);
+      CLS_LOG(20, "%s(): setting attr: %s", __func__, "cls_queue_urgent_data");
+      if (ret < 0) {
+        CLS_LOG(0, "ERROR: %s(): cls_cxx_setxattr (attr=%s) returned %d", __func__, "cls_queue_urgent_data", ret);
+        return ret;
+      }
+    }
+  }
+
+  if ((urgent_data.num_head_urgent_entries + urgent_data.num_xattr_urgent_entries) > urgent_data.num_urgent_data_entries) {
+    CLS_LOG(0, "Total num entries %u", urgent_data.num_urgent_data_entries);
+    CLS_LOG(0, "Num xattr entries %u", urgent_data.num_xattr_urgent_entries);
+    CLS_LOG(0, "Num head entries %u", urgent_data.num_head_urgent_entries);
+    CLS_LOG(0, "ERROR: Number of urgent data entries exceeded that requested by user, returning no space!");
+    return -ENOSPC;
+  }
+
+  bl_urgent_data.clear();
+  encode(urgent_data, bl_urgent_data);
+  in->clear();
+  if (! is_last_entry) {
+    cls_enqueue_op enqueue_op;
+    bufferlist bl_data;
+    encode(op.info, bl_data);
+    enqueue_op.bl_data_vec.emplace_back(bl_data);
+    CLS_LOG(1, "INFO: cls_gc_update_entry: Data size is: %u \n", bl_data.length());
+    if (has_urgent_data) {
+      enqueue_op.bl_urgent_data = bl_urgent_data;
+    }
+    encode(enqueue_op, *in);
+    ret = cls_enqueue(hctx, in, out);
+    if (ret < 0) {
+      return ret;
+    }
+  } else {
+    cls_queue_update_last_entry_op update_op;
+    encode(op.info, update_op.bl_data);
+    CLS_LOG(1, "INFO: cls_gc_update_entry: Data size is: %u \n", update_op.bl_data.length());
+    if (has_urgent_data) {
+      update_op.bl_urgent_data = bl_urgent_data;
+    }
+    encode(update_op, *in);
+    ret = cls_queue_update_last_entry(hctx, in, out);
```

Comment:
i'm not sure it's worth tracking the last_entry offset and exposing the cls_queue_get_last_entry() and cls_queue_update_last_entry() apis. while this strategy can save some queue capacity in some cases, we'd pay the extra read for cls_queue_get_last_entry() on every defer. that doesn't seem like the right trade-off. what do you think?

Reply:
If defer entry gets called frequently for the same entry, we may end up having a large number of entries for the same tag. Is that acceptable?

Reply:
first of all, i think we should impose a minimum value on `rgw_gc_obj_min_wait` using Option::set_min() to limit how frequently RGWGetObj can call cls_queue_update_last_entry(). with a sensible minimum of ~30min, we're not likely to defer many gcs, and gc processing should have no problem keeping up with the deferred entries. this will also be important to limit the size of urgent data

second, as the cluster/workload grows in scale, it becomes less and less likely that we'll see consecutive entries with the same tag. so while cls_queue_update_last_entry() is an optimization, it only applies at smaller scales where it's less important

that said, the extra read for cls_queue_get_last_entry() isn't going to kill us here; my motivation was more about simplifying the design. without this optimization, we wouldn't have to track last offset or provide the extra apis to read/write the last entry

Reply:
Ok, got it